File: django/games/objects/pong/Segment.py

from .Point import Point
from .Vector import Vector

import math

class Segment:
    
    def __init__(self, start: Point, stop: Point) -> None:
        self.start: Point = start
        self.stop: Point = stop

    def angle(self) -> float:
        return math.atan2((self.start.y - self.stop.y), (self.start.x - self.stop.x))
        
    def length(self):
        return self.start.distance(self.stop)
    
    def is_on(self, C: Point):
        return (self.start.distance(C) + self.stop.distance(C) == self.length())
    
    def __repr__(self) -> str:
        return f"Segment(start: {self.start}, stop: {self.stop})"
    
    def __str__(self) -> str:
        return f"Segment(start: {self.start}, stop: {self.stop})"
    
    def copy(self):
        return Segment(self.start.copy(), self.stop.copy())
    
    def center(self):
        return Point((self.start.x + self.stop.x) / 2, (self.start.y + self.stop.y) / 2)

    def to_dict(self) -> dict:
        
        data: dict[str: dict] = {
            "start": self.start.to_dict(),
            "stop": self.stop.to_dict(),
        }
        
        return data

File: django/frontend/static/js/api/game/pong/Segment.js
import { Point } from "./Point.js";
import { AExchangeable } from "../../AExchangable.js";
import { PongGame } from "./PongGame.js";
import { renderCube } from "../../../3D/cube.js";

class Segment extends AExchangeable
{
    /**
     * @param {Point} start
     * @param {Point} stop
     * @param {PongGame} game 
     */
    constructor(game, start = new Point(), stop = new Point())
    {
        super();
        
        /**
         * @type {Point}
         */
        this.start = start;

        /**
         * @type {Point}
         */
        this.stop = stop;

        /**
         * @type {PongGame}
         */
        this.game = game

    }

    angle()
    {
        let x = this.start.x - this.stop.x,
            y = this.start.y - this.stop.y;

        return Math.atan2(y, x);
    }

    len()
    {
        let x = this.start.x - this.stop.x,
            y = this.start.y - this.stop.y;

        return (x ** 2 + y ** 2) ** (1 / 2);
    }

    draw(ctx)
	{
		ctx.moveTo(this.start.x, this.start.y);
		ctx.lineTo(this.stop.x, this.stop.y);
	}

    /**
     * @param {[String]} additionalFieldList
     */
    export(additionalFieldList)
    {
        super.export([...additionalFieldList, "start", "stop"]);
    }
}

export { Segment }
